### DISCUSSION TITLE
How to hide label if a dataset further down on the report is empty?

### DISCUSSION THREAD

**Discussion by Didier8692** [eclipse-birt/birt#1223] [Q&A] [answered]
Hello,

I am currently implementing a report which needs to look something like this (in a PDF format, not that it matters):
LABEL "SECTION 1 TITLE"
[GRID 1 (using DataSet1)]
[GRID 2 (using DataSet2)]
[GRID 3 (using DataSet3)]
LABEL "SECTION 2 TITLE"
[GRID 4 (using DataSet4)]
[GRID 5 (using DataSet5)]
[GRID 6 (using DataSet6)]

The "SECTION 1 TITLE" label needs to be seen in the page header of all the pages of grids 1/2/3. It's a repeated header for the whole section. There are no page breaks between grids 1, 2 and 3.

Here's where my requirement gets tricky: dataSets 1, 2 and 3 could all be empty so their grids will be hidden. In that case, Section 1 would be completely empty and so I should not show the label "SECTION 1 TITLE" and the section 2 title would be the first one visible. 
This is what I am struggling to achieve. How do I hide the "SECTION 1 TITLE" label when all of the first 3 datasets are empty?
Is there a way to evaluate if datasets 1/2/3 are empty or not before BIRT starts to render their respective grids?
Is there a way to retroactively hide the section 1 label, after the grids have been evaluated?
Or is there a different approach I haven't though of?

I actually had another idea that didn't pan out:
I repeat the "SECTION 1 TITLE" label on the 3 grid headers. That way it will be shown if there's data in the grids, and hidden if the grids are empty. But this posed another problem: grid2 might start in the middle of a page right after grid1 ends. In that case, the title header will be shown in the middle of the page, but I don't want it there. I only want it at the top of each pages of the section.

I hope I'm being clear and you can help me find the best way to implement this.

Thanks,
Didier

**Comment by hvbtup** [comment, marked as answer]
The solution is based on the idea of executing the datasets you need for your decision before showing the title.

What I would do is:

Create a boolean report variable "have_section1".

For each *i* of the 3 datasets DataSet1,2,3:
Create a dynamic text item *above* section 1 title.
Bind this dynamic text item to DataSet*i*.
Use exactly the same dataset parameter bindings as in the Grid for DataSet*i*.

Note: This will normally not cause a performance issue, because BIRT caches data sets: If you use the same data set with the same parameter values, the SQL query will only be executed once.

In each of the 3 datasets, set the report variable "have_section1" to true in the onFetch event.

Use the report variable in the visibility expression of section 1 title.

**Reply by Didier8692** [reply]
Many thanks for this great answer. It fixes my problem completely.
Or almost completely. Because now my report is taking a little bit longer because after the first execution and fetching of the query for the dynamic text item, the grid takes some time as well to get the data from the cache. So the time spent in the report in queries (1 query to the data source + 1 query to the cache) is increasing by 40% compared to when I had only 1 execution. 
I can probably leave with that. But if you have a solution for this issue as well, I would be much obliged :)
